### ISSUE TITLE
Unexpected redirection of API routes with localization

### ISSUE THREAD

**Issue by rubybb** [vercel/next.js#20439]
# Bug report

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. In your next.config.js, declare under i18n the defaultLocale and other applicable supported locales.
2. Define an API route within pages/api/.
3. Visit said route.
4. It'll return a redirection code.

## Expected behavior
Next.js shouldn't redirect API routes to localized variants, at least by default.

## System information

- OS: Ubuntu 20.04.1 LTS x86_64
- Version of Next.js: v10.0.3
- Version of Node.js: v15.2.1
- Deployment: Vercel, next dev

**Comment by ijjk** [comment]
Hi, can you provide a repo with a minimal reproduction? I tried reproducing with the steps provided above but was unable to. 

Configured `next.config.js`
```js
module.exports = {
  i18n: {
    locales: ['fr', 'en'],
    defaultLocale: 'en'
  }
}
```

Next.js versions tested `v10.0.3` and `v10.0.4`

API request
```sh
curl -v 'http://localhost:3000/en/api/hello'                          
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET /en/api/hello HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< ETag: "13-5j0ZZR0tI549fSRsYxl8c9vAU78"
< Content-Length: 19
< Vary: Accept-Encoding
< Date: Thu, 24 Dec 2020 05:02:25 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< 
* Connection #0 to host localhost left intact
{"name":"John Doe"}* Closing connection 0
```

**Comment by rubybb** [comment]
> Hi, can you provide a repo with a minimal reproduction? I tried reproducing with the steps provided above but was unable to.
> 
> Configured `next.config.js`
> 
> ```js
> module.exports = {
>   i18n: {
>     locales: ['fr', 'en'],
>     defaultLocale: 'en'
>   }
> }
> ```
> 
> Next.js versions tested `v10.0.3` and `v10.0.4`
> 
> API request
> 
> ```shell
> curl -v 'http://localhost:3000/en/api/hello'                          
> *   Trying ::1...
> * TCP_NODELAY set
> * Connected to localhost (::1) port 3000 (#0)
> > GET /en/api/hello HTTP/1.1
> > Host: localhost:3000
> > User-Agent: curl/7.64.1
> > Accept: */*
> > 
> < HTTP/1.1 200 OK
> < Content-Type: application/json; charset=utf-8
> < ETag: "13-5j0ZZR0tI549fSRsYxl8c9vAU78"
> < Content-Length: 19
> < Vary: Accept-Encoding
> < Date: Thu, 24 Dec 2020 05:02:25 GMT
> < Connection: keep-alive
> < Keep-Alive: timeout=5
> < 
> * Connection #0 to host localhost left intact
> {"name":"John Doe"}* Closing connection 0
> ```

In the curl request you did, you called the localized endpoint, calling /api/hello redirects to /en/api/hello, which I'm pretty sure it shouldn't.

**Comment by ijjk** [comment]
That was on example request, I tested the other variants as well to see if there were any redirects occurring there. Note: we only apply locale redirects for the index route (/) currently so the API routes should not be affected. I added the results from the other checks below

```sh
curl -v 'http://localhost:3000/api/hello'                             
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET /api/hello HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< ETag: "13-5j0ZZR0tI549fSRsYxl8c9vAU78"
< Content-Length: 19
< Vary: Accept-Encoding
< Date: Thu, 24 Dec 2020 16:22:38 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< 
* Connection #0 to host localhost left intact
{"name":"John Doe"}* Closing connection 0
```

```sh
curl -v 'http://localhost:3000/fr/api/hello'                         
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET /fr/api/hello HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< ETag: "13-5j0ZZR0tI549fSRsYxl8c9vAU78"
< Content-Length: 19
< Vary: Accept-Encoding
< Date: Thu, 24 Dec 2020 16:22:57 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< 
* Connection #0 to host localhost left intact
{"name":"John Doe"}* Closing connection 0
```

**Comment by rubybb** [comment]
Oddly enough, I can't seem to reproduce in a fresh repository, however in the current project, It surely does redirect.

```

curl -v http://localhost:3000/api/v1/auth/login
*   Trying ::1...
* TCP_NODELAY set
* Expire in 149995 ms for 3 (transfer 0x5720b52f10)
* Expire in 200 ms for 4 (transfer 0x5720b52f10)
* Connected to localhost (::1) port 3000 (#0)
> GET /api/v1/auth/login HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.64.0
> Accept: */*
> 
< HTTP/1.1 308 Permanent Redirect
< Location: /en/api/v1/auth/login/
< Refresh: 0;url=/en/api/v1/auth/login/
< Date: Thu, 24 Dec 2020 16:49:30 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
* ```

**Comment by rubybb** [comment]
Strangely, it might be something from within my own project, I'll close this issue for now.

**Comment by ijjk** [comment]
It looks like it's applying the trailingSlash redirect and prepending the locale unexpectedly which has been fixed in the latest version of Next.js V10.0.4. If you want to avoid the redirect completely you can make sure the request does or doesn't have the trailing slash per the config in next.config.js

**Comment by rubybb** [comment]
fixed by #20751

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
